[pick](profile) Pick profile related to 3.0

zhiqiang-hhhh opened the pull request:
### What problem does this PR solve?

Pick profile related pr to branch-3.0

Related PR:
#46564 
#46720 
#46993 
#47328 
#47623 
#40185 
#48496 
#48653 



Problem Summary:

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

zhiqiang-hhhh committed:
[refactor](profile report) Remove old profile reporting. (#40185)

1. remove the old profile report
2. Add FragmentLevelProfile in new profile report.

zhiqiang-hhhh committed:
[refactor][profile] log enhancement and minor change of profile (#46564)

zhiqiang-hhhh committed:
[regression-test](profile) Add some flaky profile tests back (#46720)

zhiqiang-hhhh committed:
[fix](regression test) Fix unstable scanner_profile test (#47328)

zhiqiang-hhhh committed:
[Fix][regression-test] Fix unstable adaptive_pipeline_task_serial_read_on_limit (#46993)

adaptive_pipeline_task_serial_read_on_limit is unstable, since profile
is too much. check profile content after query finished instead of
checking after all test sql finished.

zhiqiang-hhhh committed:
[test](profile) More unit test for profile on FE (#47623)

zhiqiang-hhhh committed:
[opt](profile) Optimization for profile file format (#48496)

### What problem does this PR solve?

A new profile file format. User could decompress profile file by using
unzip directly.

```
[hezhiqiang@VM-10-2-centos log]$ unzip profile/1740745121714_33bf38e988ea4945-b585d2f74d1da3fd.zip
Archive:  profile/1740745121714_33bf38e988ea4945-b585d2f74d1da3fd.zip
  inflating: 33bf38e988ea4945-b585d2f74d1da3fd.profile
[hezhiqiang@VM-10-2-centos log]$ head 33bf38e988ea4945-b585d2f74d1da3fd.profile -n 10
�{"summaryProfile":{"counterTotalTime":{"value":0,"type":5,"level":1},"localTimePercent":0.0,"infoStrings":{"Distributed Plan":"N/A","Task Type":"QUERY","User":"root","Default Catalog":"internal","Total":"9sec745ms","Default Db":"tpch","Profile ID":"33bf38e988ea4945-b585d2f74d1da3fd","Task State":"OK","Sql Statement":"SELECT      c.c_name,      COUNT(o.o_orderkey) AS total_orders,      SUM(o.o_totalprice) AS total_spent FROM      customer c JOIN      orders o ON c.c_custkey \u003d o.o_custkey GROUP BY      c.c_name limit 20","Start Time":"2025-02-28 20:18:31","End Time":"2025-02-28 20:18:41"},"infoStringsDisplayOrder":["Profile ID","Task Type","Start Time","End Time","Total","Task State","User","Default Catalog","Default Db","Sql Statement","Distributed Plan"],"counterMap":{"TotalTime":{"value":0,"type":5,"level":1}},"childCounterMap":{},"childMap":{},"childList":[],"planNodeInfos":[],"name":"Summary","timestamp":-1,"isDone":false,"isCancel":false,"isSinkOperator":false,"nodeid":-1},"executionSummaryProfile":{"counterTotalTime":{"value":0,"type":5,"level":1},"localTimePercent":0.0,"infoStrings":{"Fragment RPC Count":"2","Analysis Time":"6ms","CreateSingleNode Time":"N/A","Get Table Version Count":"N/A","Parse SQL Time":"3ms","Get Partition Files Time":"N/A","Get Partition Version Time":"N/A","Nereids Optimize Time":"11ms","Nereids Lock Table Time":"6ms","Get Partitions Time":"N/A","Get Table Version Time":"N/A","Nereids GarbageCollect Time":"0ms","Wait and Fetch Result Time":"9sec643ms","Is Nereids":"Yes","Is Cached":"No","Workload Group":"normal","Finalize Scan Node Time":"N/A","Fetch Result Time":"9sec641ms","Instances Num Per BE":"10.16.10.2:8261:4","Plan Time":"26ms","Fragment RPC Phase1 Time":"28ms","Create Scan Range Time":"N/A","Fragment Assign Time":"3ms","Nereids BeFoldConst Time":"0ms","JoinReorder Time":"N/A","QueryDistributed Time":"N/A","Get Splits Time":"N/A","Fragment Compressed Size":"22.63 KB","Get Partition Version Count":"N/A","Fragment RPC Phase2 Time":"30ms","Schedule Time Of BE":"{\"phase1\":{\"10.16.10.2: 8261\":{\"RPC Work Time\":\"17ms\",\"RPC Latency From FE To BE\":\"10ms\",\"RPC Work Queue Time\":\"0ms\",\"RPC Latency From BE To FE\":\"1ms\"}},\"phase2\":{\"10.16.10.2: 8261\":{\"RPC Work Time\":\"4ms\",\"RPC Latency From FE To BE\":\"3ms\",\"RPC Work Queue Time\":\"0ms\",\"RPC Latency From BE To FE\":\"23ms\"}}}","Get Partition Version Count (hasData)":"N/A","Fragment Serialize Time":"9ms","Init Scan Node Time":"N/A","Trace ID":"","Nereids Rewrite Time":"10ms","Schedule Time":"70ms","Transaction Commit Time":"N/A","Nereids Translate Time":"2ms","Parallel Fragment Exec Instance Num":"48","Total Instances Num":"4","Doris Version":"e021a6a015","Nereids Distribute Time":"6ms","Nereids Analysis Time":"3ms","Write Result Time":"0ms","System Message":"N/A","Executed By Frontend":"N/A"},"infoStringsDisplayOrder":["Parse SQL Time","Nereids Lock Table Time","Nereids Analysis Time","Nereids Rewrite Time","Nereids Optimize Time","Nereids Translate Time","Nereids Distribute Time","Workload Group","Analysis Time","Plan Time","JoinReorder Time","CreateSingleNode Time","QueryDistributed Time","Init Scan Node Time","Finalize Scan Node Time","Get Splits Time","Get Partitions Time","Get Partition Files Time","Create Scan Range Time","Get Partition Version Time","Get Partition Version Count (hasData)","Get Partition Version Count","Get Table Version Time","Get Table Version Count","Schedule Time","Fragment Assign Time","Fragment Serialize Time","Fragment RPC Phase1 Time","Fragment RPC Phase2 Time","Fragment Compressed Size","Fragment RPC Count","Schedule Time Of BE","Wait and Fetch Result Time","Fetch Result Time","Write Result Time","Doris Version","Is Nereids","Is Cached","Total Instances Num","Instances Num Per BE","Parallel Fragment Exec Instance Num","Trace ID","Transaction Commit Time","System Message","Executed By Frontend","Nereids GarbageCollect Time","Nereids BeFoldConst Time"],"counterMap":{"TotalTime":{"value":0,"type":5,"level":1}},"childCounterMap":{},"childMap":{},"childList":[],"planNodeInfos":[],"name":"Execution Summary","timestamp":-1,"isDone":false,"isCancel":false,"isSinkOperator":false,"nodeid":-1},"parseSqlStartTime":1740745111965,"parseSqlFinishTime":1740745111968,"nereidsLockTableFinishTime":1740745111971,"nereidsAnalysisFinishTime":1740745111974,"nereidsRewriteFinishTime":1740745111984,"nereidsOptimizeFinishTime":1740745111995,"nereidsTranslateFinishTime":1740745111997,"nereidsGarbageCollectionTime":0,"nereidsBeFoldConstTime":0,"queryBeginTime":1740745111968,"queryAnalysisFinishTime":1740745111974,"queryJoinReorderFinishTime":-1,"queryCreateSingleNodeFinishTime":-1,"queryDistributedFinishTime":-1,"initScanNodeStartTime":-1,"initScanNodeFinishTime":-1,"finalizeScanNodeStartTime":-1,"finalizeScanNodeFinishTime":-1,"getSplitsStartTime":-1,"getPartitionsFinishTime":-1,"getPartitionFilesFinishTime":-1,"getSplitsFinishTime":-1,"createScanRangeFinishTime":-1,"queryPlanFinishTime":1740745112000,"assignFragmentTime":1740745112003,"fragmentSerializeTime":1740745112012,"fragmentSendPhase1Time":1740745112040,"fragmentSendPhase2Time":1740745112070,"fragmentCompressedSize":23173,"fragmentRpcCount":2,"queryScheduleFinishTime":1740745112070,"queryFetchResultFinishTime":1740745121713,"tempStarTime":1740745121713,"queryFetchResultConsumeTime":9641,"queryWriteResultConsumeTime":0,"getPartitionVersionTime":0,"getPartitionVersionCount":0,"getPartitionVersionByHasDataCount":0,"getTableVersionTime":0,"getTableVersionCount":0,"transactionCommitBeginTime":-1,"transactionCommitEndTime":-1,"filesystemOptTime":-1,"hmsAddPartitionTime":-1,"hmsAddPartitionCnt":0,"hmsUpdatePartitionTime":-1,"hmsUpdatePartitionCnt":0,"filesystemRenameFileCnt":0,"filesystemRenameDirCnt":0,"filesystemDeleteDirCnt":0,"filesystemDeleteFileCnt":0,"transactionType":"UNKNOWN"}�U
Changed Session Variables:
VarName                       | CurrentValue | DefaultValue
------------------------------|--------------|-------------
insert_visible_timeout_ms     | 10000        | 60000
fetch_splits_max_wait_time_ms | 4000         | 1000
exec_mem_limit                | 2147483648   | 100147483648
profile_level                 | 2            | 1
auto_profile_threshold_ms     | 1            | -1
```

Sometimes, profile could be very large, so this pr changed
implementation of `loadProfilesFromStorageIfFirstTime`.
Use a separate thread to handle profile loading after catalog is ready,
so that first loading of profile files will not block main thread. And
this thread read profile in a batch manner, max concurrency of each bath
is 10. By using batch loading, we can reduce peak of memory usage.

According to test, load of 500 profiles from disks costs 800ms(On a
x86-64 machine with 8C, HDD).
```
2025-03-02 10:11:17,019 INFO (profile-loader|195) [ProfileManager.lambda$loadProfilesFromStorageIfFirstTime$2():521] Reading 500 profiles from /mnt/datadisk1/Doris-Deploy/20250301070117-doris-master-3ea1b7c5c6/fe/log/profile
2025-03-02 10:11:17,054 INFO (profile-loader|195) [ProfileManager.lambda$loadProfilesFromStorageIfFirstTime$2():560] Processed batch 0 - 10 of 500 profiles
...
2025-03-02 10:11:17,901 INFO (profile-loader|195) [ProfileManager.lambda$loadProfilesFromStorageIfFirstTime$2():563] Load profiles into memory finished, costs 884ms
```
Total size of profile file on disk is 58MB, typical size of single test
profile file about 150KB, after decompress, its content ifs about 1.6MB.
```text
root@d1:/mnt/datadisk1/Doris-Deploy/20250301070117-doris-master-3ea1b7c5c6/fe/log# du -h profile/
58M	profile/
```
So max memory usage for the loading is about 16MB(10 * 1.6MB).

Thearas commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

zhiqiang-hhhh commented:
run buildall

zhiqiang-hhhh committed:
F

zhiqiang-hhhh commented:
run buildall

zhiqiang-hhhh committed:
NED

zhiqiang-hhhh committed:
[opt](profile) Refine process of auto_profile for less memory usage (#48653)

### What problem does this PR solve?

In previous, auto_profile_threshold is calculated by background thread
of ProfileManager.
When FE has a high qps, this implementation makes profile in memory
stacking until background thread wakes up.

So, we move the calculation of auto_profile_threshold to
Profile::update_summary.

zhiqiang-hhhh commented:
run buildall

zhiqiang-hhhh committed:
NED

zhiqiang-hhhh commented:
run buildall

zhiqiang-hhhh committed:
FX

zhiqiang-hhhh commented:
run buildall

zhiqiang-hhhh commented:
run buildall

dataroaring reviewed (approved):
LGTM

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `0.00% (0/25)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/9c27eb74d3011aefde6cf7d30db6b306b2e62c11_9c27eb74d3011aefde6cf7d30db6b306b2e62c11/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/9c27eb74d3011aefde6cf7d30db6b306b2e62c11_9c27eb74d3011aefde6cf7d30db6b306b2e62c11/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 38.87% (10165/26152) |
| Line Coverage     | 30.31% (86723/286104) |
| Region Coverage   | 29.35% (44545/151780) |
| Branch Coverage   | 25.88% (22661/87564) |
